src/engraving/rendering/score/horizontalspacing.cpp

```diff
@@ -32,17 +32,767 @@
+                continue;
+            }
+            if (nextCR->beam() != beam) {
+                continue;
+            }
+            if (thisCR->up() == nextCR->up()) {
+                return crossBeamType;
+            }
+            if (thisCR->up() && !nextCR->up()) {
+                if (thisCR->beamlet() && nextCR->beamlet()) {
+                    hasOpposingBeamlets = true;
+                }
+                upDown = true;
+            }
+            if (!thisCR->up() && nextCR->up()) {
+                downUp = true;
+            }
+            if (upDown && downUp) {
+                return crossBeamType;
+            }
+        }
+    }
+
+    crossBeamType.upDown = upDown;
+    crossBeamType.downUp = downUp;
+    crossBeamType.canBeAdjusted = canBeAdjusted;
+    crossBeamType.hasOpposingBeamlets = hasOpposingBeamlets;
+
+    return crossBeamType;
+}
+
+double HorizontalSpacing::computeMinMeasureWidth(Measure* m)
+{
+    const MStyle& style = m->style();
+    double minWidth = style.styleMM(Sid::minMeasureWidth);
+    double maxSysWidth = m->system()->width() - m->system()->leftMargin();
+
+    if (maxSysWidth <= 0) {
+        maxSysWidth = minWidth;
+    }
+    minWidth = std::min(minWidth, maxSysWidth);
+
+    if (m->ticks() < m->timesig()) {
+        minWidth *= (m->ticks() / m->timesig()).toDouble();
+    }
+
+    Segment* firstCRSegment = m->findFirstR(SegmentType::ChordRest, Fraction(0, 1));
+    if (!firstCRSegment) {
+        return minWidth;
+    }
+    if (firstCRSegment == m->firstEnabled()) {
+        return minWidth;
+    }
+
+    double startPosition = firstCRSegment->x() - firstCRSegment->minLeft();
+    if (firstCRSegment->hasAccidentals()) {
+        startPosition -= style.styleMM(Sid::barAccidentalDistance);
+    } else {
+        startPosition -= style.styleMM(Sid::barNoteDistance);
+    }
+
+    minWidth += startPosition;
+    minWidth = std::min(minWidth, maxSysWidth);
+
+    return minWidth;
+}
+
+void HorizontalSpacing::enforceMinimumMeasureWidths(const std::vector<Measure*> measureGroup)
+{
+    for (int i = 0; i < measureGroup.size(); ++i) {
+        Measure* measure = measureGroup[i];
+        double minWidth = computeMinMeasureWidth(measure);
+        double diff = minWidth - measure->width();
+        if (diff > 0) {
+            stretchMeasureToTargetWidth(measure, minWidth);
+            for (int j = i + 1; j < measureGroup.size(); ++j) {
+                measureGroup[j]->mutldata()->moveX(diff);
+            }
+        }
+    }
+}
+
+void HorizontalSpacing::stretchMeasureToTargetWidth(Measure* m, double targetWidth)
+{
+    IF_ASSERT_FAILED(targetWidth > m->width()) {
+        return;
+    }
+
+    std::vector<Spring> springs;
+    for (Segment& s : m->segments()) {
+        if (s.isChordRestType() && s.visible() && s.enabled() && !s.allElementsInvisible()) {
+            double springConst = 1 / s.stretch();
+            double width = s.width(LD_ACCESS::BAD) - s.widthOffset();

```

Comment:
What exactly is bad about it? :)

Reply:
I actually don't know 😅, it's part of some sort of sanity-checks that Igor introduced but I don't think we ended up ever fully using them (I've removed this now).